### PR TITLE
Fix use of namespace qualifiers that have not been explicitly introduced

### PR DESCRIPTION
**Pull request by morten-ofstad**
These changed are required to compile the azure-core library with VS2017 compilers. I am a bit surprised that it works with C++17 compilers, since a minimal example I made of introducing a name with a 'using X::Y::Z' declaration and then trying to refer to the name as Y::Z doesn't seem to work with any compilers (Using Compiler Explorer to test).

### CODE REVIEW

**Review by antkmsft — APPROVED** [review]
Thank you, @morten-ofstad!
The situation is similar to #4352:
"Note that we do not have CI for VS 2017, and at this point we support VS 2019 and 2022. I.e. we can't guarantee you that we won't break VS 2017 build in future commits in some other place, as we don't have tools to verify that and we don't have plans to add them. But we will likely accept fixes like this in the future as well (I hope that it won't be needed often).".

As of the PR, the CI is currently failing on clang-format in the `sdk/core/azure-core/src/http/bearer_token_authentication_policy.cpp` file.
One way to fix it, is to run "`find ./sdk \( -iname '*.hpp' -o -iname '*.cpp' \) ! -iname 'json.hpp' -exec clang-format -i {} \;`" but if your dev environment is not really set up for that, let me know, we can figure something out - for example I could run it on my machine, and propose a suggestion to this PR.

File: sdk/core/azure-core/src/http/bearer_token_authentication_policy.cpp [hunk 1]
```diff
@@ -13,6 +13,7 @@ using Azure::Core::Context;
 using Azure::Core::Credentials::AuthenticationException;
 using Azure::Core::Credentials::TokenRequestContext;
 using Azure::Core::Credentials::_detail::AuthorizationChallengeHelper;
+using Azure::Core::Http::Request;
```
Comment (ahsonkhan) [inline sdk/core/azure-core/src/http/bearer_token_authentication_policy.cpp:16]: Is this line necessary? If so, why?

**Review by ahsonkhan — APPROVED** [review]

File: sdk/core/azure-core/src/http/bearer_token_authentication_policy.cpp (hunk 1 above)
Comment (morten-ofstad) [reply sdk/core/azure-core/src/http/bearer_token_authentication_policy.cpp:16]: It's using that definition e.g. in the signature of  BearerTokenAuthenticationPolicy::AuthenticateAndAuthorizeRequest():

> void BearerTokenAuthenticationPolicy::AuthenticateAndAuthorizeRequest(
>     Request& request,  <---- here
>     Credentials::TokenRequestContext const& tokenRequestContext,
>     TokenRequestContext const& tokenRequestContext,
>     Context const& context) const

Without it I get these errors:

> C:\Project\open-vds\3rdparty\azure-sdk-for-cpp-1.10.3\sdk\core\azure-core\src\http\bearer_token_authentication_policy.cpp(20): error C2065: 'Request': undeclared identifier
> C:\Project\open-vds\3rdparty\azure-sdk-for-cpp-1.10.3\sdk\core\azure-core\src\http\bearer_token_authentication_policy.cpp(20): error C2065: 'request': undeclared identifier
> C:\Project\open-vds\3rdparty\azure-sdk-for-cpp-1.10.3\sdk\core\azure-core\src\http\bearer_token_authentication_policy.cpp(21): error C2275: 'Azure::Core::Http::Policies::NextHttpPolicy': illegal use of this type as an expression
>   C:\Project\open-vds\3rdparty\azure-sdk-for-cpp-1.10.3\sdk\core\azure-core\inc\azure/core/http/policies/policy.hpp(303): note: see declaration of 'Azure::Core::Http::Policies::NextHttpPolicy'
> C:\Project\open-vds\3rdparty\azure-sdk-for-cpp-1.10.3\sdk\core\azure-core\src\http\bearer_token_authentication_policy.cpp(21): error C2146: syntax error: missing ')' before identifier 'nextPolicy'
> C:\Project\open-vds\3rdparty\azure-sdk-for-cpp-1.10.3\sdk\core\azure-core\src\http\bearer_token_authentication_policy.cpp(23): error C2143: syntax error: missing ';' before '{'
> C:\Project\open-vds\3rdparty\azure-sdk-for-cpp-1.10.3\sdk\core\azure-core\src\http\bearer_token_authentication_policy.cpp(23): error C2447: '{': missing function header (old-style formal list?)
> C:\Project\open-vds\3rdparty\azure-sdk-for-cpp-1.10.3\sdk\core\azure-core\src\http\bearer_token_authentication_policy.cpp(44): error C2065: 'Request': undeclared identifier
> C:\Project\open-vds\3rdparty\azure-sdk-for-cpp-1.10.3\sdk\core\azure-core\src\http\bearer_token_authentication_policy.cpp(44): error C2065: 'request': undeclared identifier
> C:\Project\open-vds\3rdparty\azure-sdk-for-cpp-1.10.3\sdk\core\azure-core\src\http\bearer_token_authentication_policy.cpp(45): error C2065: 'nextPolicy': undeclared identifier
> C:\Project\open-vds\3rdparty\azure-sdk-for-cpp-1.10.3\sdk\core\azure-core\src\http\bearer_token_authentication_policy.cpp(45): error C2275: 'Azure::Core::Http::Policies::NextHttpPolicy': illegal use of this type as an expression
>   C:\Project\open-vds\3rdparty\azure-sdk-for-cpp-1.10.3\sdk\core\azure-core\inc\azure/core/http/policies/policy.hpp(303): note: see declaration of 'Azure::Core::Http::Policies::NextHttpPolicy'
> C:\Project\open-vds\3rdparty\azure-sdk-for-cpp-1.10.3\sdk\core\azure-core\src\http\bearer_token_authentication_policy.cpp(46): error C2275: 'Azure::Core::Context': illegal use of this type as an expression
>   C:\Project\open-vds\3rdparty\azure-sdk-for-cpp-1.10.3\sdk\core\azure-core\inc\azure/core/context.hpp(46): note: see declaration of 'Azure::Core::Context'
> C:\Project\open-vds\3rdparty\azure-sdk-for-cpp-1.10.3\sdk\core\azure-core\src\http\bearer_token_authentication_policy.cpp(46): error C2059: syntax error: 'const'
> C:\Project\open-vds\3rdparty\azure-sdk-for-cpp-1.10.3\sdk\core\azure-core\src\http\bearer_token_authentication_policy.cpp(47): error C2143: syntax error: missing ';' before '{'
> C:\Project\open-vds\3rdparty\azure-sdk-for-cpp-1.10.3\sdk\core\azure-core\src\http\bearer_token_authentication_policy.cpp(47): error C2447: '{': missing function header (old-style formal list?)
> C:\Project\open-vds\3rdparty\azure-sdk-for-cpp-1.10.3\sdk\core\azure-core\src\http\bearer_token_authentication_policy.cpp(54): error C2061: syntax error: identifier 'Request'
> C:\Project\open-vds\3rdparty\azure-sdk-for-cpp-1.10.3\sdk\core\azure-core\src\http\bearer_token_authentication_policy.cpp(56): error C2511: 'bool Azure::Core::Http::Policies::_internal::BearerTokenAuthenticationPolicy::AuthorizeRequestOnChallenge(const std::string &) const': overloaded member function not found in 'Azure::Core::Http::Policies::_internal::BearerTokenAuthenticationPolicy'
>   C:\Project\open-vds\3rdparty\azure-sdk-for-cpp-1.10.3\sdk\core\azure-core\inc\azure/core/http/policies/policy.hpp(546): note: see declaration of 'Azure::Core::Http::Policies::_internal::BearerTokenAuthenticationPolicy'
> C:\Project\open-vds\3rdparty\azure-sdk-for-cpp-1.10.3\sdk\core\azure-core\src\http\bearer_token_authentication_policy.cpp(58): error C2065: 'request': undeclared identifier
> C:\Project\open-vds\3rdparty\azure-sdk-for-cpp-1.10.3\sdk\core\azure-core\src\http\bearer_token_authentication_policy.cpp(59): error C2065: 'context': undeclared identifier
> C:\Project\open-vds\3rdparty\azure-sdk-for-cpp-1.10.3\sdk\core\azure-core\src\http\bearer_token_authentication_policy.cpp(65): error C2065: 'Request': undeclared identifier
> C:\Project\open-vds\3rdparty\azure-sdk-for-cpp-1.10.3\sdk\core\azure-core\src\http\bearer_token_authentication_policy.cpp(65): error C2065: 'request': undeclared identifier
> C:\Project\open-vds\3rdparty\azure-sdk-for-cpp-1.10.3\sdk\core\azure-core\src\http\bearer_token_authentication_policy.cpp(66): error C2275: 'Azure::Core::Credentials::TokenRequestContext': illegal use of this type as an expression
>   C:\Project\open-vds\3rdparty\azure-sdk-for-cpp-1.10.3\sdk\core\azure-core\inc\azure/core/credentials/credentials.hpp(44): note: see declaration of 'Azure::Core::Credentials::TokenRequestContext'
> C:\Project\open-vds\3rdparty\azure-sdk-for-cpp-1.10.3\sdk\core\azure-core\src\http\bearer_token_authentication_policy.cpp(66): error C2059: syntax error: 'const'
> C:\Project\open-vds\3rdparty\azure-sdk-for-cpp-1.10.3\sdk\core\azure-core\src\http\bearer_token_authentication_policy.cpp(68): error C2143: syntax error: missing ';' before '{'
> C:\Project\open-vds\3rdparty\azure-sdk-for-cpp-1.10.3\sdk\core\azure-core\src\http\bearer_token_authentication_policy.cpp(68): error C2447: '{': missing function header (old-style formal list?)
>

**Comment by morten-ofstad** [comment]
> Thank you, @morten-ofstad! The situation is similar to #4352: "Note that we do not have CI for VS 2017, and at this point we support VS 2019 and 2022. I.e. we can't guarantee you that we won't break VS 2017 build in future commits in some other place, as we don't have tools to verify that and we don't have plans to add them. But we will likely accept fixes like this in the future as well (I hope that it won't be needed often).".
> 
> As of the PR, the CI is currently failing on clang-format in the `sdk/core/azure-core/src/http/bearer_token_authentication_policy.cpp` file. One way to fix it, is to run "`find ./sdk \( -iname '*.hpp' -o -iname '*.cpp' \) ! -iname 'json.hpp' -exec clang-format -i {} \;`" but if your dev environment is not really set up for that, let me know, we can figure something out - for example I could run it on my machine, and propose a suggestion to this PR.

I did not make any changes to the formatting, and it's not that easy for me to run clang-format, so I would appreciate your help with this.

**Comment by morten-ofstad** [comment]
> These changed are required to compile the azure-core library with VS2017 compilers. I am a bit surprised that it works with C++17 compilers, since a minimal example I made of introducing a name with a 'using X::Y::Z' declaration and then trying to refer to the name as Y::Z doesn't seem to work with any compilers (Using Compiler Explorer to test).

... I looked a bit more into this, and it is because the rules for unqualified name lookup was changed in C++17 so it also searches the nested namespaces where the function/method was first declared when resolving names (of types) in the function definition (including the parameter list). So your code is perfectly valid C++17 but not valid C++14.

I worked with Jørgen who contributed the #4352 changes, this time I was doing the upgrade to a more recent version of azure-sdk-for-cpp and since you wrote that you "will likely accept fixes like this in the future as well" I decided to make a PR with the patch I had to apply in our build system...

**Comment by LarryOsterman** [comment]
> > These changed are required to compile the azure-core library with VS2017 compilers. I am a bit surprised that it works with C++17 compilers, since a minimal example I made of introducing a name with a 'using X::Y::Z' declaration and then trying to refer to the name as Y::Z doesn't seem to work with any compilers (Using Compiler Explorer to test).
> 
> ... I looked a bit more into this, and it is because the rules for unqualified name lookup was changed in C++17 so it also searches the nested namespaces where the function/method was first declared when resolving names (of types) in the function definition (including the parameter list). So your code is perfectly valid C++17 but not valid C++14.
> 
> I worked with Jørgen who contributed the #4352 changes, this time I was doing the upgrade to a more recent version of azure-sdk-for-cpp and since you wrote that you "will likely accept fixes like this in the future as well" I decided to make a PR with the patch I had to apply in our build system...

That's fascinating, because it implies that there's a bug in MSVC, clang and GCC - when they added C++17 name resolution support, they didn't include the change in an "if C++17" conditional, and thus they accidentally changed the rules for C++ namespace resolution even when /std::c++14 is specified. That's the kind of mistake I could easily imagine being made. 

And of course it would only be discovered when using an older compiler that hadn't been updated to the new rules.

I'll start a discussion on our internal C++ alias to see if this break can be fixed in MSVC.

**Comment by LarryOsterman** [comment]
> > > These changed are required to compile the azure-core library with VS2017 compilers. I am a bit surprised that it works with C++17 compilers, since a minimal example I made of introducing a name with a 'using X::Y::Z' declaration and then trying to refer to the name as Y::Z doesn't seem to work with any compilers (Using Compiler Explorer to test).
> > 
> > 
> > ... I looked a bit more into this, and it is because the rules for unqualified name lookup was changed in C++17 so it also searches the nested namespaces where the function/method was first declared when resolving names (of types) in the function definition (including the parameter list). So your code is perfectly valid C++17 but not valid C++14.
> > I worked with Jørgen who contributed the #4352 changes, this time I was doing the upgrade to a more recent version of azure-sdk-for-cpp and since you wrote that you "will likely accept fixes like this in the future as well" I decided to make a PR with the patch I had to apply in our build system...
> 
> That's fascinating, because it implies that there's a bug in MSVC, clang and GCC - when they added C++17 name resolution support, they didn't include the change in an "if C++17" conditional, and thus they accidentally changed the rules for C++ namespace resolution even when /std::c++14 is specified. That's the kind of mistake I could easily imagine being made.
> 
> And of course it would only be discovered when using an older compiler that hadn't been updated to the new rules.
> 
> I'll start a discussion on our internal C++ alias to see if this break can be fixed in MSVC.

@morten-ofstad I asked our C++ standards committee members and they're not aware of any changes in type resolution introduced in C++17. Do you happen to have the DR number which changed the behavior, or a reference page describing the type resolution change?

**Comment by morten-ofstad** [comment]
> @morten-ofstad I asked our C++ standards committee members and they're not aware of any changes in type resolution introduced in C++17. Do you happen to have the DR number which changed the behavior, or a reference page describing the type resolution change?

Sorry, I think I was confused by the nested-namespace support that was added in C++17. In fact the unqualified name lookup rules that require searching enclosing namespaces of the class definition were already part of C++11 ( 3.4.1 Unqualified name lookup, section 6), but it appears it was not implemented correctly in Visual Studio prior to Visual Studio 2019 (where nested-namespace support was added).

**Comment by LarryOsterman** [comment]
> > @morten-ofstad I asked our C++ standards committee members and they're not aware of any changes in type resolution introduced in C++17. Do you happen to have the DR number which changed the behavior, or a reference page describing the type resolution change?
> 
> Sorry, I think I was confused by the nested-namespace support that was added in C++17. In fact the unqualified name lookup rules that require searching enclosing namespaces of the class definition were already part of C++11 ( 3.4.1 Unqualified name lookup, section 6), but it appears it was not implemented correctly in Visual Studio prior to Visual Studio 2019 (where nested-namespace support was added).

Ah, that makes much more sense. So this PR is to work around a bug in VS 2017, which isn't at all surprising.

**Pull request merged**